File: frontend/src/api/identityClient.ts
import { request } from "./client";
import { boolean, nullable, record, string, uuid } from "./dto";

function hash(value: unknown) {
  const result = string(value);
  if (!/^[a-f0-9]{64}$/.test(result)) throw new Error("Invalid identity hash");
  return result;
}
function number(value: unknown) {
  if (typeof value !== "number" || !Number.isSafeInteger(value) || value < 0) throw new Error("Invalid identity number");
  return value;
}
function list<T>(value: unknown, limit: number, parse: (item: unknown) => T): T[] {
  if (!Array.isArray(value) || value.length > limit) throw new Error("Invalid identity list");
  return value.map(parse);
}
function context(value: unknown) {
  const data = record(value);
  return { materialId: uuid(data.material_id), identity: string(data.technical_identity), folder: nullable(data.folder_path),
    brandId: uuid(data.published_brand_id), category: string(data.main_category_code), number: number(data.sequence_number),
    name: string(data.material_name) };
}
function finding(value: unknown) {
  const item = record(value);
  const code = string(item.code);
  if (!/^[A-Z][A-Z0-9_]{0,99}$/.test(code)) throw new Error("Invalid identity finding");
  return { code, path: string(item.path) };
}
function plan(value: unknown) {
  const data = record(value); const worker = record(data.worker_plan); const metadata = record(worker.metadata);
  if (worker.schema_version !== 1 || worker.planner_version !== "identity-plan-1") throw new Error("Unknown identity planner");
  const errors = list(worker.errors, 40010, finding);
  const ready = boolean(worker.ready);
  if (ready !== (errors.length === 0)) throw new Error("Inconsistent identity plan");
  const source = context(data.source_context); const target = context(data.target_context);
  if (source.materialId !== target.materialId || source.folder !== worker.source_path || target.folder !== worker.target_path) throw new Error("Mismatched identity plan");
  return { source, target, generation: number(data.generation), hash: hash(data.proposal_hash), reservesNumber: boolean(data.reserves_number),
    workerHash: hash(worker.plan_hash), sourceHash: hash(worker.source_revision_hash), ready, errors,
    warnings: list(worker.warnings, 10, finding),
    metadata: { beforeHash: metadata.before_hash === null ? null : hash(metadata.before_hash), afterHash: metadata.after_hash === null ? null : hash(metadata.after_hash),
      fields: list(metadata.changed_fields, 9, string) },
    changes: list(worker.changes, 20000, (value) => { const item = record(value);
      if (item.kind !== "file" && item.kind !== "directory") throw new Error("Unknown rename kind");
      return { source: string(item.source), target: string(item.target), kind: item.kind,
        hash: item.sha256 === null ? null : hash(item.sha256) };
    }),
  };
}
export type IdentityPlan = ReturnType<typeof plan>;
export interface IdentityTarget { target_brand_id: string; main_category_code: string; target_parent: string; }
export interface IdentityConfirmation extends IdentityTarget {
  idempotency_key: string; expected_generation: number; expected_proposal_hash: string; reason: string; warnings_acknowledged: boolean;
}
function operation(value: unknown) {
  const data = record(value); const status = string(data.status);
  if (!["RUNNING", "COMPLETED", "ROLLED_BACK", "RECOVERY_REQUIRED", "REJECTED"].includes(status)) throw new Error("Unknown identity outcome");
  const result = data.result === null ? null : record(data.result);
  return { id: uuid(data.id), status, source: context(data.source_context), target: context(data.target_context),
    reason: string(data.reason), createdAt: string(data.created_at), actorId: uuid(data.actor_id),
    failure: result ? nullable(result.failure_code) : null };
}
export type IdentityOperation = ReturnType<typeof operation>;
export const identityClient = {
  async operations(id: string) {
    const data = record(await request(`/materials/${uuid(id)}/identity-operations`));
    return { enabled: boolean(data.mutations_enabled), items: list(data.operations, 100, operation) };
  },
  async plan(id: string, target: IdentityTarget) { return plan(await request(`/materials/${uuid(id)}/identity-plan`, "POST", target)); },
  async confirm(id: string, payload: IdentityConfirmation) { return operation(await request(`/materials/${uuid(id)}/identity-confirm`, "POST", payload)); },
  async resume(id: string, operationId: string) { return operation(await request(`/materials/${uuid(id)}/identity-operations/${uuid(operationId)}/resume`, "POST")); },
};

File: backend/app/identity_client.py
"""Strict private-worker identity protocol. Responses never contain source bytes."""
import json
from typing import Annotated, Literal, Protocol, Self
from uuid import UUID

import httpx
from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.inventory_client import INVENTORY_CODES, WorkerInventoryClient, validate_relative_path
from app.material_review import canonical_hash
from app.schemas import FolderPath, Sha256

IDENTITY_CODES = INVENTORY_CODES | frozenset({
    "IDENTITY_CROSS_DEVICE_UNSUPPORTED", "IDENTITY_PARENT_ENTRY_LIMIT", "IDENTITY_PATH_LIMIT",
    "IDENTITY_SOURCE_UNSUPPORTED", "IDENTITY_TARGET_COLLISION", "IDENTITY_TARGET_INSIDE_SOURCE", "IDENTITY_TARGET_INVALID",
    "METADATA_REWRITE_UNSUPPORTED", "METADATA_UNMAPPED_REFERENCE", "SOURCE_METADATA_MISSING",
    "SOURCE_METADATA_UNSAFE_FILE", "SOURCE_METADATA_UNREADABLE", "SOURCE_METADATA_TOO_LARGE",
    "IDENTITY_ENTRY_CHANGED", "IDENTITY_METADATA_CHANGED", "IDENTITY_OPERATION_FAILED", "IDENTITY_OUTPUT_CHANGED",
    "IDENTITY_PLAN_BLOCKED", "IDENTITY_PLAN_CHANGED", "IDENTITY_PREPARATION_INTERRUPTED", "IDENTITY_PREPARATION_REJECTED",
    "IDENTITY_RECOVERY_REQUIRED", "JOURNAL_BUSY", "JOURNAL_INVALID_STATE", "JOURNAL_REQUEST_CONFLICT",
    "JOURNAL_ROOT_OVERLAPS_SOURCE", "JOURNAL_TARGET_EXISTS", "JOURNAL_WRITE_FAILED", "JOURNAL_BACKUP_FAILED",
    "JOURNAL_ID_INVALID", "JOURNAL_NAME_INVALID", "JOURNAL_PLATFORM_UNSUPPORTED", "JOURNAL_RENAME_FAILED",
    "JOURNAL_SIZE_LIMIT", "JOURNAL_UNSAFE_STORAGE", "SOURCE_MUTATIONS_DISABLED", "IDENTITY_UNAVAILABLE",
})
MAX_IDENTITY_RESPONSE_BYTES = 32 * 1024 * 1024
METADATA_FIELDS = {"FOLDER", "MANUFACTURER", "PRODUCT_NAME", "CATEGORY", "PRODUCT_NUMBER", "BASE_NAME",
                   "TEXTURE_SIZE_SOURCE", "COLOR.measured_from", "SOURCE.SBS"}


class IdentityClientError(RuntimeError):
    def __init__(self, code="IDENTITY_UNAVAILABLE"):
        self.code = code if code in IDENTITY_CODES else "IDENTITY_UNAVAILABLE"
        super().__init__("Identity operation could not be verified.")


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)


class IdentityFinding(StrictModel):
    code: Annotated[str, Field(max_length=100)]
    path: FolderPath

    @model_validator(mode="after")
    def check(self) -> Self:
        if self.code not in IDENTITY_CODES: raise ValueError("Unknown identity finding")
        validate_relative_path(self.path)
        return self


class IdentityChange(StrictModel):
    kind: Literal["file", "directory"]
    source: FolderPath
    target: FolderPath
    sha256: Sha256 | None

    @model_validator(mode="after")
    def check(self) -> Self:
        validate_relative_path(self.source); validate_relative_path(self.target)
        if self.source == self.target or (self.kind == "file") != (self.sha256 is not None):
            raise ValueError("Invalid rename")
        return self


class IdentityMetadata(StrictModel):
    before_hash: Sha256 | None
    after_hash: Sha256 | None
    changed_fields: Annotated[list[str], Field(max_length=9)]

    @model_validator(mode="after")
    def check(self) -> Self:
        if self.changed_fields != sorted(set(self.changed_fields)) or set(self.changed_fields) - METADATA_FIELDS:
            raise ValueError("Invalid metadata field list")
        if self.changed_fields and (self.before_hash is None or self.after_hash is None or self.before_hash == self.after_hash):
            raise ValueError("Missing metadata proof")
        return self


class IdentityPlan(StrictModel):
    schema_version: Literal[1]
    planner_version: Literal["identity-plan-1"]
    source_path: FolderPath
    target_path: FolderPath
    source_revision_hash: Sha256
    changes: Annotated[list[IdentityChange], Field(max_length=20_000)]
    metadata: IdentityMetadata
    errors: Annotated[list[IdentityFinding], Field(max_length=40_010)]
    warnings: Annotated[list[IdentityFinding], Field(max_length=10)]
    ready: bool
    plan_hash: Sha256

    @model_validator(mode="after")
    def check(self) -> Self:
        validate_relative_path(self.source_path); validate_relative_path(self.target_path)
        if self.ready != (not self.errors) or self.source_path == self.target_path:
            raise ValueError("Invalid plan outcome")
        sources = [item.source for item in self.changes]
        if sources != sorted(set(sources)): raise ValueError("Duplicate or unsorted renames")
        old = self.source_path.rsplit("/", 1)[-1]; new = self.target_path.rsplit("/", 1)[-1]
        for item in self.changes:
            mapped = "/".join(new + part[len(old):] if part == old or part.startswith((old + "_", old + ".")) else part
                              for part in item.source.split("/"))
            if item.target != mapped: raise ValueError("Unexpected rename")
        if self.ready and self.metadata.before_hash is not None:
            if self.metadata.after_hash is None or bool(self.metadata.changed_fields) != (self.metadata.before_hash != self.metadata.after_hash):
                raise ValueError("Incomplete metadata proof")
        return self

    def verify_request(self, request):
        if (self.source_path, self.target_path) != (request["folder_path"], request["target_path"]):
            raise ValueError("Unexpected identity target")
        digest = canonical_hash({"plan": self.model_dump(mode="json", exclude={"plan_hash"}),
                                 "brand_name": request["brand_name"], "material_name": request["material_name"]})
        if self.plan_hash != digest: raise ValueError("Identity plan hash mismatch")


class IdentityResult(StrictModel):
    operation_id: UUID
    status: Literal["COMPLETED", "ROLLED_BACK", "RECOVERY_REQUIRED", "REJECTED"]
    plan_hash: Sha256
    source_path: FolderPath
    target_path: FolderPath
    source_revision_hash: Sha256 | None
    target_revision_hash: Sha256 | None
    failure_code: str | None

    @model_validator(mode="after")
    def check(self) -> Self:
        if self.failure_code is not None and self.failure_code not in IDENTITY_CODES:
            raise ValueError("Unknown operation finding")
        if (self.status == "COMPLETED") != (self.target_revision_hash is not None and self.failure_code is None):
            raise ValueError("Contradictory operation result")
        if self.status != "COMPLETED" and (self.target_revision_hash is not None or self.failure_code is None):
            raise ValueError("Unexpected failed operation result")
        if (self.status == "REJECTED") != (self.source_revision_hash is None):
            raise ValueError("Missing source proof")
        return self


class IdentityClient(Protocol):
    def plan(self, request: dict) -> IdentityPlan: ...
    def execute(self, request: dict) -> IdentityResult: ...


class WorkerIdentityClient(WorkerInventoryClient):
    def __init__(self, base_url, token=None, enabled=False, timeout_seconds=300):
        super().__init__(base_url, timeout_seconds)
        self.base_url = base_url.rstrip("/")
        self._token = token
        self.enabled = enabled

    def _request(self, path, payload, *, mutation=False):
        if mutation and (not self.enabled or not self._token): raise IdentityClientError("SOURCE_MUTATIONS_DISABLED")
        headers = {"Authorization": "Bearer " + self._token.get_secret_value()} if mutation else {}
        try:
            with httpx.stream("POST", self.base_url + path, json=payload, headers=headers,
                              timeout=self.timeout, trust_env=False, follow_redirects=False) as response:
                declared = response.headers.get("Content-Length")
                if declared is not None and not 0 <= int(declared) <= MAX_IDENTITY_RESPONSE_BYTES: raise ValueError()
                chunks = []; size = 0
                for chunk in response.iter_bytes(chunk_size=65536):
                    size += len(chunk)
                    if size > MAX_IDENTITY_RESPONSE_BYTES: raise ValueError()
                    chunks.append(chunk)
                content = b"".join(chunks)
                if response.status_code != 200:
                    if response.status_code in {401, 409, 422, 503}:
                        body = json.loads(content)
                        raise IdentityClientError(body.get("detail", {}).get("code"))
                    raise IdentityClientError()
                return content
        except IdentityClientError: raise
        except (httpx.HTTPError, ValueError, TypeError, AttributeError, ArithmeticError):
            raise IdentityClientError() from None

    def plan(self, request):
        try:
            result = IdentityPlan.model_validate_json(self._request("/internal/material-identity-plan", request))
            result.verify_request(request)
            return result
        except IdentityClientError: raise
        except (ValueError, TypeError, AttributeError, ArithmeticError): raise IdentityClientError() from None

    def execute(self, request):
        try:
            result = IdentityResult.model_validate_json(self._request("/internal/material-identity-execute", request, mutation=True))
            if (str(result.operation_id), result.plan_hash, result.source_path, result.target_path) != (
                    request["operation_id"], request["expected_plan_hash"], request["folder_path"], request["target_path"]):
                raise ValueError("Unexpected operation result")
            return result
        except IdentityClientError: raise
        except (ValueError, TypeError, AttributeError, ArithmeticError): raise IdentityClientError() from None
